### ISSUE TITLE
QUESTION: Best way to work with migrations in 2020

### ISSUE THREAD

**Issue by mrtnzagustin** [balderdashy/sails#6949]
Hi there. Im working with sails since the last six months and at this moment i am looking for a good way to work with migrations and database schema updates for new releases of all my backends.
The question is, wich tools do the sailsjs team recommends to work with? i was trying to catch the queries that auto-migrations with 'alter' do, but that queries are not logged (not at silly level).
The other two tools (sails-migrations and sails-db-migrate ) are no longer maintained. 
In the docs, knexjs is named (this tool is external and is currently maintained) but no examples of how implement the tool with sails are given. Sequalize for example has a migration tool. I was looking just for something like doctrine:schema:update (this tools check doctrine php entities and outputs the differences with the database in sql sentences). 
A basic idea i have is to make a new database with the new version models, and use a external tool that show diffs with the production database and output that sql sentences.
I am very happy with waterline but i need some guide with this phase. 
Thanks in advance.

**Comment by sailsbot** [comment]
@mrtnzagustin Thanks for posting!  We'll take a look as soon as possible.

In the mean time, there are a few ways you can help speed things along:
 - look for a workaround.  _(Even if it's just temporary, sharing your solution can save someone else a lot of time and effort.)_
 - tell us why this issue is important to you and your team.  What are you trying to accomplish?  _(Submissions with a little bit of human context tend to be easier to understand and faster to resolve.)_
 - make sure you've provided clear instructions on how to reproduce the bug from a clean install.
 - double-check that you've provided all of the requested version and dependency information.  _(Some of this info might seem irrelevant at first, like which database adapter you're using, but we ask that you include it anyway.  Oftentimes an issue is caused by a confluence of unexpected factors, and it can save everybody a ton of time to know all the details up front.)_
 - read the [code of conduct](https://sailsjs.com/documentation/contributing/code-of-conduct).
 - if appropriate, ask your business to [sponsor your issue](https://sailsjs.com/support).   _(Open source is our passion, and our core maintainers volunteer many of their nights and weekends working on Sails.  But you only get so many nights and weekends in life, and stuff gets done a lot faster when you can work on it during normal daylight hours.)_
 - let us know if you are using a 3rd party plugin; whether that's a database adapter, a non-standard view engine, or any other dependency maintained by someone other than our core team.  _(Besides the name of the 3rd party package, it helps to include the exact version you're using.  If you're unsure, check out [this list of all the core packages we maintain](https://sailsjs.com/architecture).)_ 
<hr/>

Please remember: never post in a public forum if you believe you've found a genuine security vulnerability.  Instead, [disclose it responsibly](https://sailsjs.com/security).

For help with questions about Sails, [click here](http://sailsjs.com/support).

**Comment by mrtnzagustin** [comment]
After i notice sails dont have a migrations tool included, i did a small analysis of the two most used competitors: sequalize and typeORM, plus knexjs.

1) SEQUELIZE
- Migrations: Sequelize have an integrated cli for migrations but do not generate an output file with auto-migrations. So you must first know de api and after that elaborate step by step each migration. Plus: there is a sequelize auto migrations project but it has a few pull request pending, not up to date: https://github.com/flexxnn/sequelize-auto-migrations
- ORM: Just compatible with sql (waterline also supports mongo). In the other functionalities, sequelize do the same things as waterline. 
- Compatibility with sails: Sequelize have a mantained sails-hook: https://github.com/KSDaemon/sails-hook-sequelize and a non maintened sails-hook-sequelize-blueprints https://github.com/cesardeazevedo/sails-hook-sequelize-blueprints

2) TYPEORM
- Migrations: TypeORM have an integrated cli for migrations and it do generates auto-migrations, but only for typescript. The typeorm cli (that includes migrations and auto-migrations) works only with typescript.
- ORM and compatibilty with sails: In generally, the creators says that it does have support for es5, es6 and es7, i did test the orm and make a little hook to make it work (actually it works) but all the documentation is mayoritary for typescript. So you can make it work without the cli but you have to work harder if you not like typescript (i dont like the idea of use sails with ts just for using typeorm). If you want just the orm and not the cli, you can make it work.

3) KNEXJS: 
- Is a tool that dont generate auto-migrations. So, like sequelize, you have to know in deep the api to make migrations files manually using specific functions like  `knex.schema.createTable('users', function (table) {  table.increments();  table.string('name'); table.timestamps();})`. There is a repo that add the functionality to define a json schema that you can use for create schemas from cero: https://github.com/danielgindi/knex-schema-builder (i did not test the auto-migration posibilitie with this one yet).
- Little discussion about auto-migration and schema file support: https://github.com/knex/knex/issues/1086

My conclussion: If you use sails, at january 2020, best use it with waterline and if you want, and you really needs, add a tool like knex to do a more specific job. You will have to mantain your models and also your specific knex migrations files (problem). Mi decission is to use sails and i addes a script that creates a blank schema in a new database, and compare that database with (using dbdiff) the production databases. The tool outputs all the differences in sentences you can use to make the update, probably adding some control to prevent data lost. If i can make knex works with schema and auto generation files, i will use that with waterline orm. 


PD: Sorry for my english.
PD2: I'll love to know the core team opinion on this analysis.

**Comment by grantbi** [comment]
I would also like to know the best approach to this.

**Comment by mrtnzagustin** [comment]
@johnabrams7 any chance to give us a little feedback on this? thank you all

**Comment by madisonhicks** [comment]
Hi, @mrtnzagustin—

Thanks for taking the time to do this write-up!

Our preferred way of handling database migrations is to do them manually, using hand-written scripts. This has been our modus operandi for a couple of years now, and it's the best solution we've found thus far!

**Comment by NachtRitter** [comment]
@madisonhicks 
Modern ORM without migrations and seeds/fixtures is a useless thing. Waterline is the poorest ORM that I've even used. For the core team would be better to switch to smth like Sequelize or TypeORM instead of reinventing the wheel. Those ORMs have much more reach functionality and big community support.
It's a pity to see how good framework just has been dying slowly.
In nowadays of classes, modules and TypeScript you are looking like an old dinosaur from 90's.

**Comment by mikermcneil** [comment]
@NachtRitter Thanks for sharing your opinion.  Sails is built based on the best practices my team and I have found to be effective for our projects, from startups to the Fortune 500.  We continue to iterate on those best practices over time, but we have to balance trendiness versus value when evaluating features like ES classes, ES modules, and tighter TypeScript integration.

As of now, after keeping a close eye on them for a few years, I am still not convinced that those features add enough value compared to the complexity they introduce.

@mrtnzagustin We added [a bit more color on this subject](https://github.com/balderdashy/sails-docs/pull/1303/files) in the docs as a result of you bringing this up.  Thanks again!
